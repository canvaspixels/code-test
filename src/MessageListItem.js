import React, { Component } from 'react';
import PropTypes from 'prop-types';

export default class MessageListItem extends Component {
  state = {
    showUser: false
  }

  showUser = (userId) => {
    this.setState({ showUser: true })
  }

  hideUser = (userId) => {
    this.setState({ showUser: false })
  }

  formatDate(timestamp) {
    const date = new Date(timestamp);

    return `${date.getDate()}/${date.getMonth() + 1}/${date.getFullYear()}`
  }

  render() {
    const { message } = this.props;
    return (
      <li>
        <a
          href=""
          onMouseEnter={this.showUser}
          onMouseLeave={this.hideUser}
        >
          <img src={message.user.avatar} alt=""/>

          <div>{message.message}</div>
          <time dateTime={message.timestamp}>{this.formatDate(message.timestamp)}</time>
        </a>

        { this.state.showUser && (
          <div>{message.user.email}</div>
        )}
      </li>
    );
  }
};

MessageListItem.propTypes = {
  message: PropTypes.shape({
    id: PropTypes.string,
    userId: PropTypes.string,
    message: PropTypes.string,
    timestamp: PropTypes.string,
    user: PropTypes.shape({
      id: PropTypes.string,
      firstName: PropTypes.string,
      lastName: PropTypes.string,
      email: PropTypes.string,
      avatar: PropTypes.string,
      ip: PropTypes.string,
    })
  })
};

export const messageListItemProps = MessageListItem.propTypes.message;