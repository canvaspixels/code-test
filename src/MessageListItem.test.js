import React from 'react';
import { shallow } from 'enzyme';
import MessageListItem from './MessageListItem';

const getStubMessage = () => ({
  id: 'cd445e6d-e514-424f-ba8f-16ec842002c6',
  userId: 'fe27b760-a915-475c-80bb-7cdf14cc6ef3',
  message: 'Quisque erat eros, viverra eget, congue eget, semper rutrum, nulla.',
  timestamp: '2017-02-09T04:27:38Z',
  user: {
    id: 'fe27b760-a915-475c-80bb-7cdf14cc6ef3',
    firstName: 'Albert',
    lastName: 'Rose',
    email: 'arosec@bbb.org',
    avatar: 'http://dummyimage.com/100x100.jpg/5fa2dd/ffffff',
    ip: '20.79.231.223'
  }
});

describe('MessageListItem', () => {
  it('matches snapshot', () => {
    const component = shallow(<MessageListItem message={getStubMessage()} />);

    expect(component).toMatchSnapshot();
  });

  it('matches snapshot with userâ€™s email shown', () => {
    const component = shallow(<MessageListItem message={getStubMessage()} />);

    component.instance().showUser();
    component.update();

    expect(component).toMatchSnapshot();
  });

  describe('showUser()', () => {
    it('sets showUser on the state to true', () => {
      const component = shallow(<MessageListItem message={getStubMessage()} />);
      const instance = component.instance();
      instance.setState = jest.fn();
      instance.showUser();

      expect(instance.setState).toHaveBeenCalledWith({ showUser: true });
    });
  });

  describe('hideUser()', () => {
    it('sets showUser on the state to false', () => {
      const component = shallow(<MessageListItem message={getStubMessage()} />);
      const instance = component.instance();
      instance.setState = jest.fn();
      instance.hideUser();

      expect(instance.setState).toHaveBeenCalledWith({ showUser: false });
    });
  });

  describe('formatDate()', () => {
    it('sets showUser on the state to false', () => {
      const component = shallow(<MessageListItem message={getStubMessage()} />);
      const instance = component.instance();

      expect(instance.formatDate('2017-02-09T04:27:38Z')).toBe('9/2/2017');
      expect(instance.formatDate('2018-05-31T04:27:38Z')).toBe('31/5/2018');
    });
  });
});