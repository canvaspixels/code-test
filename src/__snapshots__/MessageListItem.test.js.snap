exports[`MessageListItem matches snapshot 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <li>
    <a
        href=""
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}>
        <img
            alt=""
            src="http://dummyimage.com/100x100.jpg/5fa2dd/ffffff" />
        <div>
            Quisque erat eros, viverra eget, congue eget, semper rutrum, nulla.
        </div>
        <time
            dateTime="2017-02-09T04:27:38Z">
            9/2/2017
        </time>
    </a>
</li>,
  "nodes": Array [
    <li>
      <a
            href=""
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}>
            <img
                  alt=""
                  src="http://dummyimage.com/100x100.jpg/5fa2dd/ffffff" />
            <div>
                  Quisque erat eros, viverra eget, congue eget, semper rutrum, nulla.
            </div>
            <time
                  dateTime="2017-02-09T04:27:38Z">
                  9/2/2017
            </time>
      </a>
</li>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <MessageListItem
        message={
                Object {
                        "id": "cd445e6d-e514-424f-ba8f-16ec842002c6",
                        "message": "Quisque erat eros, viverra eget, congue eget, semper rutrum, nulla.",
                        "timestamp": "2017-02-09T04:27:38Z",
                        "user": Object {
                          "avatar": "http://dummyimage.com/100x100.jpg/5fa2dd/ffffff",
                          "email": "arosec@bbb.org",
                          "firstName": "Albert",
                          "id": "fe27b760-a915-475c-80bb-7cdf14cc6ef3",
                          "ip": "20.79.231.223",
                          "lastName": "Rose",
                        },
                        "userId": "fe27b760-a915-475c-80bb-7cdf14cc6ef3",
                      }
        } />,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": MessageListItem {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "hideUser": [Function],
        "props": Object {
          "message": Object {
            "id": "cd445e6d-e514-424f-ba8f-16ec842002c6",
            "message": "Quisque erat eros, viverra eget, congue eget, semper rutrum, nulla.",
            "timestamp": "2017-02-09T04:27:38Z",
            "user": Object {
              "avatar": "http://dummyimage.com/100x100.jpg/5fa2dd/ffffff",
              "email": "arosec@bbb.org",
              "firstName": "Albert",
              "id": "fe27b760-a915-475c-80bb-7cdf14cc6ef3",
              "ip": "20.79.231.223",
              "lastName": "Rose",
            },
            "userId": "fe27b760-a915-475c-80bb-7cdf14cc6ef3",
          },
        },
        "refs": Object {},
        "showUser": [Function],
        "state": Object {
          "showUser": false,
        },
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <li>
          <a
                    href=""
                    onMouseEnter={[Function]}
                    onMouseLeave={[Function]}>
                    <img
                              alt=""
                              src="http://dummyimage.com/100x100.jpg/5fa2dd/ffffff" />
                    <div>
                              Quisque erat eros, viverra eget, congue eget, semper rutrum, nulla.
                    </div>
                    <time
                              dateTime="2017-02-09T04:27:38Z">
                              9/2/2017
                    </time>
          </a>
</li>,
        "_debugID": 2,
        "_renderedOutput": <li>
          <a
                    href=""
                    onMouseEnter={[Function]}
                    onMouseLeave={[Function]}>
                    <img
                              alt=""
                              src="http://dummyimage.com/100x100.jpg/5fa2dd/ffffff" />
                    <div>
                              Quisque erat eros, viverra eget, congue eget, semper rutrum, nulla.
                    </div>
                    <time
                              dateTime="2017-02-09T04:27:38Z">
                              9/2/2017
                    </time>
          </a>
</li>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <MessageListItem
    message={
        Object {
            "id": "cd445e6d-e514-424f-ba8f-16ec842002c6",
            "message": "Quisque erat eros, viverra eget, congue eget, semper rutrum, nulla.",
            "timestamp": "2017-02-09T04:27:38Z",
            "user": Object {
              "avatar": "http://dummyimage.com/100x100.jpg/5fa2dd/ffffff",
              "email": "arosec@bbb.org",
              "firstName": "Albert",
              "id": "fe27b760-a915-475c-80bb-7cdf14cc6ef3",
              "ip": "20.79.231.223",
              "lastName": "Rose",
            },
            "userId": "fe27b760-a915-475c-80bb-7cdf14cc6ef3",
          }
    } />,
}
`;

exports[`MessageListItem matches snapshot with userâ€™s email shown 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <li>
    <a
        href=""
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}>
        <img
            alt=""
            src="http://dummyimage.com/100x100.jpg/5fa2dd/ffffff" />
        <div>
            Quisque erat eros, viverra eget, congue eget, semper rutrum, nulla.
        </div>
        <time
            dateTime="2017-02-09T04:27:38Z">
            9/2/2017
        </time>
    </a>
    <div>
        arosec@bbb.org
    </div>
</li>,
  "nodes": Array [
    <li>
      <a
            href=""
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}>
            <img
                  alt=""
                  src="http://dummyimage.com/100x100.jpg/5fa2dd/ffffff" />
            <div>
                  Quisque erat eros, viverra eget, congue eget, semper rutrum, nulla.
            </div>
            <time
                  dateTime="2017-02-09T04:27:38Z">
                  9/2/2017
            </time>
      </a>
      <div>
            arosec@bbb.org
      </div>
</li>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <MessageListItem
        message={
                Object {
                        "id": "cd445e6d-e514-424f-ba8f-16ec842002c6",
                        "message": "Quisque erat eros, viverra eget, congue eget, semper rutrum, nulla.",
                        "timestamp": "2017-02-09T04:27:38Z",
                        "user": Object {
                          "avatar": "http://dummyimage.com/100x100.jpg/5fa2dd/ffffff",
                          "email": "arosec@bbb.org",
                          "firstName": "Albert",
                          "id": "fe27b760-a915-475c-80bb-7cdf14cc6ef3",
                          "ip": "20.79.231.223",
                          "lastName": "Rose",
                        },
                        "userId": "fe27b760-a915-475c-80bb-7cdf14cc6ef3",
                      }
        } />,
      "_debugID": 3,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": MessageListItem {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "hideUser": [Function],
        "props": Object {
          "message": Object {
            "id": "cd445e6d-e514-424f-ba8f-16ec842002c6",
            "message": "Quisque erat eros, viverra eget, congue eget, semper rutrum, nulla.",
            "timestamp": "2017-02-09T04:27:38Z",
            "user": Object {
              "avatar": "http://dummyimage.com/100x100.jpg/5fa2dd/ffffff",
              "email": "arosec@bbb.org",
              "firstName": "Albert",
              "id": "fe27b760-a915-475c-80bb-7cdf14cc6ef3",
              "ip": "20.79.231.223",
              "lastName": "Rose",
            },
            "userId": "fe27b760-a915-475c-80bb-7cdf14cc6ef3",
          },
        },
        "refs": Object {},
        "showUser": [Function],
        "state": Object {
          "showUser": true,
        },
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 2,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <li>
          <a
                    href=""
                    onMouseEnter={[Function]}
                    onMouseLeave={[Function]}>
                    <img
                              alt=""
                              src="http://dummyimage.com/100x100.jpg/5fa2dd/ffffff" />
                    <div>
                              Quisque erat eros, viverra eget, congue eget, semper rutrum, nulla.
                    </div>
                    <time
                              dateTime="2017-02-09T04:27:38Z">
                              9/2/2017
                    </time>
          </a>
          <div>
                    arosec@bbb.org
          </div>
</li>,
        "_debugID": 4,
        "_renderedOutput": <li>
          <a
                    href=""
                    onMouseEnter={[Function]}
                    onMouseLeave={[Function]}>
                    <img
                              alt=""
                              src="http://dummyimage.com/100x100.jpg/5fa2dd/ffffff" />
                    <div>
                              Quisque erat eros, viverra eget, congue eget, semper rutrum, nulla.
                    </div>
                    <time
                              dateTime="2017-02-09T04:27:38Z">
                              9/2/2017
                    </time>
          </a>
          <div>
                    arosec@bbb.org
          </div>
</li>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <MessageListItem
    message={
        Object {
            "id": "cd445e6d-e514-424f-ba8f-16ec842002c6",
            "message": "Quisque erat eros, viverra eget, congue eget, semper rutrum, nulla.",
            "timestamp": "2017-02-09T04:27:38Z",
            "user": Object {
              "avatar": "http://dummyimage.com/100x100.jpg/5fa2dd/ffffff",
              "email": "arosec@bbb.org",
              "firstName": "Albert",
              "id": "fe27b760-a915-475c-80bb-7cdf14cc6ef3",
              "ip": "20.79.231.223",
              "lastName": "Rose",
            },
            "userId": "fe27b760-a915-475c-80bb-7cdf14cc6ef3",
          }
    } />,
}
`;
